server/resource/caldav_service.go

```diff
@@ -1,18 +1,545 @@
+	"time"
 )
 
-func NewCaldavServer(CaldavListenInterface string)*http.Server{
+var Schema = `
+CREATE EXTENSION IF NOT EXISTS pgcrypto;
+
+DROP TABLE IF EXISTS calendar;
+DROP TABLE IF EXISTS users;
+DROP TABLE IF EXISTS collection;
+DROP TABLE IF EXISTS collection_role;
+
+
+CREATE TABLE calendar (
+	-- id         UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+	id         SERIAL PRIMARY KEY,
+	owner_id   INT, /* NOT NULL, */
+	created    TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+	modified   TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+	collection INT,
+	rpath      TEXT,
+	content    TEXT
+);
+
+CREATE TABLE collection (
+	-- id          UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+	id          SERIAL PRIMARY KEY,
+	owner_id    INT,
+	name        VARCHAR(64),
+	description TEXT,
+);
+
+DROP TYPE IF EXISTS perm;
+CREATE TYPE perm AS ENUM ('admin', 'write', 'read', 'none');
+
+CREATE TABLE collection_role (
+	collection_id   INT,
+	user_id         INT,
+	permission perm DEFAULT 'none',
+);
+
+CREATE TABLE users (
+	-- id         UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+	id         SERIAL PRIMARY KEY,
+	username   TEXT UNIQUE NOT NULL,
+	email      TEXT UNIQUE NOT NULL,
+	password   VARCHAR(64) NOT NULL, /* crypt('input', password) */
+	firstname  TEXT,
+	lastname   TEXT,
+);
+
+CREATE INDEX user_index ON users (username, password);
+CREATE INDEX user_cal_index ON users (id, firstname, lastname);
+CREATE INDEX cal_index ON calendar (rpath);
+CREATE INDEX cal_user_index ON calendar (rpath, owner_id);`
+
+type CalDavStorage struct {
+	cruds database.DatabaseConnection
+	UserID int64
+	Mailer Mail
+	User string
+	Email string
+}
+
+type DummyResourceAdapter struct{
+	resourcePath string
+	resourceData string
+}
+
+
+
+func (cs *CalDavStorage) GetResourcesByList(rpaths []string) ([]data.Resource, error) {
+	results := []data.Resource{}
+
+	for _, rpath := range rpaths {
+		resource, found, err := cs.GetShallowResource(rpath)
+
+		if err != nil && err != errs.ResourceNotFoundError {
+			return nil, err
+		}
+
+		if found {
+			results = append(results, *resource)
+		}
+	}
+
+	return results, nil
+}
+
+func (cs *CalDavStorage) GetShallowResource(rpath string) (*data.Resource, bool, error) {
+	resources, err := cs.GetResources(rpath, false)
+
+	if err != nil {
+		return nil, false, err
+	}
+
+	if resources == nil || len(resources) == 0 {
+		return nil, false, errs.ResourceNotFoundError
+	}
+
+	res := resources[0]
+	return &res, true, nil
+}
+
+func (cs *CalDavStorage) GetResources(rpath string, withChildren bool) ([]data.Resource, error) {
+	result := []data.Resource{}
+
+	a, err := cs.haveAccess(rpath, "read")
+	if err != nil {
+		log.Error(err, "failed to get Access [" + rpath + "]")
+		return nil, err
+	}
+
+	if ! a {
+		log.Info("no access to collection [" + rpath + "]")
+		return nil, nil
+	}
+
+	var rows *sql.Rows
+	rows, err = cs.cruds.Query("SELECT rpath FROM calendar WHERE rpath = $1 AND owner_id = $2 ", rpath, cs.UserID)
```

Comment:
queries to be built using goqu so that it works across sqlite/mysql/postgres. the $1 placeholder will break.